ntoskrnl/io/pnpmgr/pnpirp.c

```diff
@@ -306,3 +306,61 @@ PiIrpQueryPnPDeviceState(
 
     return status;
 }
+
+// IRP_MN_QUERY_CAPABILITIES (0x09)
+NTSTATUS
+PiIrpQueryPnPDeviceCapabilities(
+    _In_ PDEVICE_NODE DeviceNode,
+    _Out_ PDEVICE_CAPABILITIES DeviceCaps)
+{
+    PAGED_CODE();
+
+    ASSERT(DeviceNode);
+
+    RtlZeroMemory(DeviceCaps, sizeof(DEVICE_CAPABILITIES));
+    DeviceCaps->Size = sizeof(DEVICE_CAPABILITIES);
+    DeviceCaps->Version = 1;
+    DeviceCaps->Address = -1;
+    DeviceCaps->UINumber = -1;
+
+    IO_STACK_LOCATION stack = {
+        .MajorFunction = IRP_MJ_PNP,
+        .MinorFunction = IRP_MN_QUERY_CAPABILITIES
+    };
+    stack.Parameters.DeviceCapabilities.Capabilities = DeviceCaps;
+
+    PVOID Dummy;
+    return IopSynchronousCall(DeviceNode->PhysicalDeviceObject, &stack, &Dummy);
+}
+
+// IRP_MN_QUERY_ID (0x13)
+NTSTATUS
+PiIrpQueryPnPDeviceId(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ BUS_QUERY_ID_TYPE IdType,
+    _Out_ PWCHAR *Id)
+{
+    PAGED_CODE();
+
+    ASSERT(DeviceNode);
+    ASSERT(IdType == BusQueryDeviceID || IdType == BusQueryHardwareIDs ||
+           IdType == BusQueryCompatibleIDs || IdType == BusQueryInstanceID ||
+           IdType == BusQueryDeviceSerialNumber || IdType == BusQueryContainerID);
+
+    IO_STACK_LOCATION stack = {
+        .MajorFunction = IRP_MJ_PNP,
+        .MinorFunction = IRP_MN_QUERY_ID
```

Comment:
Do not initialize this struct in C99-style! It is not compatible to older compilers.